spec/ics-2-consensus-requirements/README.md

```diff
@@ -0,0 +1,76 @@
+
+In order to achive this, a blockchain should be able to check 1. whether the block is valid and 2. whether the packet is included in the block or not. In this proposal, we introduce an abstract concept of blockchains, called `Header`. `Header` requires its consensus algorithm to satisfy some properties, including deterministic safety, lightclient compatibility and valid transition of the state machine.
+
+## Specification
+
+### Motivation
+
+`Header` defines required properties of the blockchain on the network. The implementors can check whether the consensus that they are using is qualified to be connected to the network or not. If not, they can modify the algorithm or wrap it with additional logic to make it compatible with the specification.
+
+### Desired Properties
+
+`Header` is `(p : Maybe<Header>, v : Header -> bool, s : []byte -> []byte)` where `p` is parent block, `v` is lightclient verifier and `s` is state. The other parts of the protocol, such as connections and channels, are defined over `Header`s, so if a blockchain wants to establish an IBC communication with another, it is required to satisfy the properties of `Header`. 
+
+Definitions:
+
+1. Direct children have their parent as `p` and are verifiable with their parent.
+2. The height of a block is the step of referring `p` required to get to `nil`
+
+Requirements:
+
+1. Headers have only one direct child
+If a blockchain has deterministic safety(as opposed to probabilistic safety in Nakamoto consensus), then there cannot be more than one child for each block. In Tendermint, this assumption breaks when +1/3 validators are malicious, producing multiple blocks those are direct children of a single block and all of them are verifiable with the parent. It makes conflicting packets delivered from the failed chain, so for the chains who are receiving from this chain, **fraud proof #10** mechanism should be applied in this case. However, he failed chain itself also need to be recovered and reconnected again. It will be covered in **byzantine recovery strategies #6**.
+
+2. Headers have at least one direct child
+There is at least one direct child for all blocks, meaning that the lightclient logic can proceed the blocks one by one even in the worst case. If it not satisfied then there can be a point where the lightclient stops and cannot proceed, which halts IBC connection unexpectedly. This also can be violated when the blocks are restarted out of the consensus, for example in Tendermint, (1/2 < validators < 2/3) forked out the chain. This also will be covered in **byzantine recovery strategies**.
+
+3. If a block verifies another block then it is a descendant of the block
+Lightclient should not verify packets which are not in its chain. If the verifier returns true for a block that is not a descendant, it simply means that there is an error in the lightclient logic.
+
+4. Header can have a state only if it is a valid transition from the parent's(see the next paragraph)
```

Comment:
This is not a property of the consensus logic, but rather the application logic on the chain, correct?
Tendermint could run great and enforce 1-3, but a buggy abci app may violate 4.

Maybe this needs to be addressed at another level (eg. this is a valid state from a valid child of a trusted parent, *but* the data contained is not "proper").

Reply:
Since now we removed the `Connection` and `Channel` from the consensus requirements we can remove this too. The `KVStore` definition should be placed here.